Allow TableStyle to handle row headers

Xitian9 opened the pull request:
This addresses some of the points in https://github.com/muesli4/table-layout/issues/20#issuecomment-1060164646. In particular, row headers are now rendered in the table, along with special line styles available within the row header (as there are in the column headers).

I have chosen to display a row header only in the first line of a row group, thus breaking the symmetry between row groups and row headers, so that question has also been resolved.

Also simplifies the creation of new TableStyle from TableStyleSpec, allowing it to handle ASCII tables as well, and creating an even simpler constructor simpleTableStyleSpec.

Xitian9 commented:
Here is an overly fancy example from the test suite. It renders better in my terminal.

```
╭───╥───────────────────┬───────────────────┬────────────┰────────────┬────────────┬────────────┬────────────┰────────────┬────────────╮
│   ║     Some text     │   Some numbers    │     X      ┃     Z      │     W      │     A      │     B      ┃    Text    │     Y      │
╞═══╬═══════════════════╪═══════════════════╪════════════╋════════════╧════════════╪════════════╧════════════╋════════════╧════════════╡
│ 1 ║ This is long text ┆    4.20000000     ┆    foo     ┃    blah         bloo    ┆    blop         blog    ┃   Short         baz     │
│   ║       Short       ┆ 200300400500600.2 ┆    bar     ┃   yadda         yoda    ┆   yeeda         york    ┃   Short        wibble   │
├───╫┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄╂┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┼┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄╂┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┤
│ 2 ║ This is long text ┆    4.20000000     ┆    foo     ┃  bibbidy      babbidy   ┆    boo          blue    ┃   Short        wobble   │
╰───╨───────────────────┴───────────────────┴────────────┸─────────────────────────┴─────────────────────────┸─────────────────────────╯
```

muesli4 commented:
The end-result looks great, well done.  I will have a closer look soon.  (Just did some of the simple ones first.)

muesli4 commented on src/Text/Layout/Table.hs:
```diff
@@ -245,12 +244,12 @@ checkeredCells f g = zipWith altLines $ cycle [[f, g], [g, f]]
-colsG :: Monoid a => [Position V] -> [Col a] -> RowGroup a
+colsG :: Cell a => [Position V] -> [Col a] -> RowGroup a
 colsG ps = rowsG . colsAsRows ps
 
 -- | Create a 'RowGroup' by aligning the columns vertically. Each column uses
 -- the same vertical positioning.
-colsAllG :: Monoid a => Position V -> [Col a] -> RowGroup a
+colsAllG :: Cell a => Position V -> [Col a] -> RowGroup a
 colsAllG p = rowsG . colsAsRowsAll p
```
Why is this necessary?

Xitian9 replied:
Not all things you would want to put into a table are instances of `Monoid`, and this would preclude their use with this function. In particular, the proposed `ElidableList` in one of the draft PRs does not have a `Monoid` instance.

The only thing the monoid constraint is needed for is padding, for which `mempty` is used. Under the assumption that one would only use this to later be rendered, we would already have a `Cell` instance and so may as well use the new `emptyCell` element of the typeclass.

That said, the most general solution may be to take the padding as an argument: `colsGWith :: a -> [Position V] -> [Col a] -> RowGroup a`. We could leave the existing `colsG = colsGWith mempty`, but I'm not sure this would be worth the effort.

Xitian9 replied:
Just a follow-up that this ‘most general solution’ is actually implemented already as `vPad` and `vPadAll`. These functions are just convenience calls to that.

muesli4 replied:
I am happy with your initial explanation already.

Xitian9 commented:
Anything I can do to help this along?

muesli4 commented on src/Text/Layout/Table/Cell.hs:
```diff
@@ -37,17 +37,22 @@ class Cell a where
 
+    -- | Create an empty cell.
+    -- This should satisfy `buildCell emptyCell = mempty`.
+    emptyCell :: a
```
I'm not sure if I asked this before but what is the point of it?  If this is just there to be used once in an internal algorithm then I have to question its existence.

Xitian9 replied:
It's needed for padding, the same as the Monoid constraint on `colsG` and friends. Whenever we have more than one `Row` in a `RowGroup`, we display the row header only for the first row, and need to pad out the remaining rows with something. That is the role that `emptyCell` fills.

Without something to put there we would not be able to create row headers, at least not without doing something else, like repeating the row headers for every row in the group, which is not desired.

Xitian9 replied:
Another thing to note: `Cell` is pretty much defined as being a typeclass which maps to various `Monoid`, though `buildCell` (since `StringBuilder` is a subclass of `Monoid`). In this context, `emptyCell` has the fairly natural and lawful definition of being an element which maps to `mempty` under `buildCell`.

muesli4 replied:
Thanks!

muesli4 commented on src/Text/Layout/Table/Cell/WideString.hs:
```diff
@@ -31,7 +31,7 @@ instance Cell WideString where
 -- The provided `Bool` determines whether to continue dropping zero-width
 -- characters after the requested width has been dropped.
 dropWide :: Bool -> Int -> String -> String
-dropWide _ i [] = []
+dropWide _ _ [] = []
```
Unrelated change.

muesli4 commented on src/Text/Layout/Table/LineStyle.hs:
```diff
@@ -100,13 +102,67 @@ asciiVertical NoLine     = ""
+roundedAsciiJoinString4 w      e      NoLine NoLine | w == e = asciiHorizontal w
+roundedAsciiJoinString4 w      e      n      s               = arJoins w e n s
+
+-- | Draw Ascii line joins with pluses. Arguments are in the order west, east,
+-- north, south.
+--
+-- Only 'NoLine', 'SingleLine', and 'DoubleLine' are supported by Ascii. Other
+-- line styles are treated as 'SingleLine'.
+aJoins :: LineStyle -> LineStyle -> LineStyle -> LineStyle -> String
+aJoins _ _ DoubleLine _          = "++"
+aJoins _ _ _          DoubleLine = "++"
+aJoins _ _ _          _          = "+"
+
+-- | Draw Ascii line joins with rounded joins. Arguments are in order west,
+-- east, north, south.
+--
+-- Only 'NoLine', 'SingleLine', and 'DoubleLine' are supported by Ascii. Other
+-- line styles are treated as 'SingleLine'.
+arJoins :: LineStyle -> LineStyle -> LineStyle -> LineStyle -> String
+-- Top joins
+arJoins _          _          NoLine     DoubleLine = ".."
+arJoins _          _          NoLine     _          = "."
+-- Bottom joins
+arJoins _          _          DoubleLine NoLine     = "''"
+arJoins _          _          _          NoLine     = "'"
+-- Left joins
+arJoins NoLine     DoubleLine DoubleLine _          = "::"
+arJoins NoLine     DoubleLine _          _          = ":"
+arJoins NoLine     SingleLine DoubleLine _          = "++"
+arJoins NoLine     SingleLine _          _          = "+"
+-- Right joins
+arJoins DoubleLine NoLine     DoubleLine _          = "::"
+arJoins DoubleLine NoLine     _          _          = ":"
+arJoins SingleLine NoLine     DoubleLine _          = "++"
+arJoins SingleLine NoLine     _          _          = "+"
+-- Interior joins
+arJoins DoubleLine _          DoubleLine _          = "::"
+arJoins DoubleLine _          _          _          = ":"
+arJoins _          _          DoubleLine _          = "++"
+arJoins _          _          _          _          = "+"
```
If this can be a separate commit functionally then please make it one.

Xitian9 replied:
I've separated this code in particular into a separate commit. It's not used at that point, but it doesn't affect compilation or functionality.

muesli4 reviewed (commented):
In general, it seems that there are some unrelated changes.  Please separate as far as possible (i.e. without breaking functionality).  Enormous commits also slow down review considerably.

If I am completely wrong with my judgment then please let me know.

Xitian9 committed:
Introduce functions for constructing ASCII joins

Xitian9 committed:
Allow TableStyle to handle row headers

This adds emptyCell as a part of the Cell specification, a way of
constructing empty cells. This allows Cell to replace the Monoid
constraint in colsG, colsAllG, colsAsRows, and colsAsRowsAll.

This also simplifies the creation of new TableStyle from TableStyleSpec,
allowing it to handle ASCII tables as well, and creating an even simpler
constructor simpleTableStyleSpec.

Xitian9 committed:
Clean up -Wall warnings

muesli4 commented on src/Text/Layout/Table/LineStyle.hs:
```diff
@@ -100,13 +102,67 @@ asciiVertical NoLine     = ""
-asciiJoinString _          _          = "+"
+asciiJoinString h v = asciiJoinString4 h h v v
+
+-- | Ascii representations for joins using rounded joins.
```
Please use the term `ASCII`.

muesli4 commented on src/Text/Layout/Table/Style.hs:
```diff
@@ -3,69 +3,139 @@
+--
+-- This is quite low-level and difficult to construct by hand. If you want to
+-- construct your own, you may wish to use the higher-level interface provided
+-- by (in increasing order of detail): 'simpleTableStyleSpec', 'TableStyleSpec',
+-- 'unicodeTableStyleFromSpec', 'asciiTableStyleFromSpec', 'tableStyleFromSpec'.
```
Do bullet points make sense?

Xitian9 replied:
Yes they do.

muesli4 reviewed (commented):
Awesome.  I like how your abstraction simplifies the style definitions.
